Wiki side of slim expressions

tcnh opened the pull request:
Looking at the acceptance test page, I'd say this covers all cases where symbols are set on the Slim runner side.

I don't see a solution yet on using expressions on symbols that are only set on the executor side (i.e. maps set by map of maps fixture) and I am starting to doubt there is one, given the strict separation of these two

six42 committed:
sync with unclebob master

six42 committed:
Merge branch 'upstream/master'

six42 committed:
Revert "sync with unclebob master"

This reverts commit 79412ba694a23f2df5afc0e4ee4540956aebf778.

six42 committed:
Merge pull request #6 from unclebob/master

sync with unclebob:master

six42 committed:
Merge branch 'unclebob/master'

jruijgers committed:
859: Change behavior for loading classes by graceful name

When the class is not found, try to load the class by changing the case
of the first letter of the classname. This helps fallback scenarios
where the Disgracer class does not give the correct name.

I.E.:
|Table:rest|...|
|...|

This will look for a class 'rest', while we want to find the class
'Rest'.

jruijgers committed:
859: Change behavor for finding methods by their graceful name

When matching the method by name, if the name not matches also try to
match with the first character changed in case.

I.E.:
|Is Match|...|

This will look for a method 'IsMatch', while we want to find the method
'isMatch'.

fixes #859, fixes #863, fixes #655, fixes #783

jruijgers committed:
Issue 859: improve changed behavior

First try the old logic to find the class/method by the graceful name,
then try finding it with the first letter changed in case.

jruijgers committed:
Issue 859: update documentation

Commit:
Evaluate TodaysDate and expresions inside Preformatted blocks

pepeZpew committed:
fit.RowFixture replace ConcurrentHashMap with plain HashMap to handle null keys

jaruzafa committed:
Added jacoco plugin

jaruzafa committed:
Moved plugin declaration

six42 committed:
Merge remote-tracking branch 'upstream/master'

NicoleRauch committed:
The SuiteSpecificationRunner was also a TraversalListener. Let's have two different objects for these different purposes.

NicoleRauch committed:
No need to have a field here as the list is only being used inside one method.

NicoleRauch committed:
Even the test class served the double purpose of being a TraversalListener as well.

NicoleRauch committed:
WikiPageFinder was a finder and a traverser at the same time. Now we have two classes for this.

NicoleRauch committed:
There is no need for an interface and an implementation that are both in the same directory.

NicoleRauch committed:
The TraverseDirective also was a TraversalListener. Now we have two classes.

NicoleRauch committed:
Better reuse constructors across the same class.

amolenaar committed:
Upgrade Gradle to 4.3

This should help in building a Java 9 version of FitNesse.

fhoeben committed:
Allow custom scenario subclasses, and tables called in a "scenario's context" to access the scenario table

fhoeben committed:
Allow tables defined in other packages to see ScenarioTestContext class

fhoeben committed:
Remove unneeded import

fhoeben committed:
Only fail test run if too few tests are executed, just log on too many

Commit:
Unified all tests in this directory to use the HitCollector.

Commit:
Converted the last test that was also a TraversalListener.

Commit:
Pulled the hidden ReferenceRenamingTraverser out of the ReferenceRenamer.

Commit:
Pulled the TraversalListener out of the SearchReplaceResponder.

Commit:
Extracted the traversers into their own files.

Commit:
Minor cleanup (renamed variable etc.)

Commit:
In order to exclude skipped pages AND their children from a suite specification run, we made the PageCrawler (that is used by the SuiteSpecificationRunner) skip those pages.

Commit:
We want to configure the traversal of the page crawler, not the creation.

amolenaar committed:
Move all calls to Class.forName to a central place

... so I can deal with plugin loading in a decent way.

amolenaar committed:
Use a dedicated classLoader for plugins, instead of (ab)using the system class loader

The plugin class loader creates a special ClassLoader. This class loader is used (by means of static assignment) on all places where previously a `Class.forName` was used.

amolenaar committed:
Pass classloader as instance variable where possible

Reducing the dependency on the static ClassUtils methods.
Ideally ClassUtils is a temporary solution (I know...).

amolenaar committed:
Clean up unused instantiation variables

amolenaar committed:
Update XML page serialization to avoid extra whitespace when run with Java 9

amolenaar committed:
Remove set and map related queries from HybridDecisionTable

The order of sets and maps is undefined, therefore we can not safely assume an order.

amolenaar committed:
Build Java 9 on Travis

amolenaar committed:
Merge pull request #1116 from amolenaar/gradle-4

Upgrade Gradle to 4.3

fhoeben committed:
Ensure classloader used by default can be changed via ClassUtils

fhoeben committed:
Ensure classloader used by default context configuration can be changed via ClassUtils

fhoeben committed:
Single return when getting classloader

fhoeben committed:
Remove no longer valid todo

fhoeben committed:
Restructure plugins class loader

fhoeben committed:
Rename pluginClassLoader to PluginClassLoaderFactory as it is not the classloader itself but delivers one

amolenaar committed:
Do no longer change java.class.path with plugin path

amolenaar committed:
PluginsClassLoader only has static members

Make constructor private to reflect this.

amolenaar committed:
Merge pull request #5 from fhoeben/java-9-classloader-take-2

Java 9 classloader take 2

amolenaar committed:
Keep FitNesse configuration extensible in JUnit runner

amolenaar committed:
Add warning around use of ClassUtils

amolenaar committed:
Move SSL parameter creation our of Ssl*SocketFactory

This way I can load the SslParameters via the plugins class loader from the wiki side. Also the interface is more explicit/properly typed.

amolenaar committed:
Pass class loader to in-process test systems

This allows test runners and test system specific classes such as FixtureInteraction's to be loaded from plugins as well.

amolenaar committed:
Set context class loader for in-process slim service to the plugin class loader

amolenaar committed:
Set context class loader for in-process FIT service to the plugin class loader

amolenaar committed:
Update release notes

amolenaar committed:
Merge pull request #1107 from NicoleRauch/suiteSpecIgnoresChildrenToo

A SuiteQuery must neither run test pages that are excluded nor children of excluded pages

six42 committed:
The slim client process should be stopped if the startup didn't success

six42 committed:
Allow text before the Slim Header Message

This is required if the remote_debug function is used.
It can also be helpful if you do integration testing and your application is writing extra messages before the Slim Client is started.

six42 committed:
Speedup unit testing

by removing tests which are run during acceptance testing

six42 committed:
Fixed #1059

six42 committed:
Upgraded test class to removed depreciation

six42 committed:
Renamed page to better describe behavior

amolenaar committed:
Merge pull request #1103 from fhoeben/allowScenarioTableAccessViaScenarioContext

Allow scenario table access via scenario context

amolenaar committed:
Merge pull request #1118 from six42/port1_with_debug

 Allow text before the Slim Header Message

amolenaar committed:
Merge pull request #1105 from fhoeben/dontFailRunWhenTooManyTestsCompleted

Only fail test run if too few tests are executed

amolenaar committed:
Merge pull request #1117 from amolenaar/java-9-classloader-take-2

Update FitNesse for Java 9

amolenaar committed:
Include all wiki files under version control

FitNesseRoot is used for tests on the code base. Therefore there is a chance files are added to a distribution that are not under version control. This change ensures wiki files are only added if under version control.

This fixes #1112 in a different way.

tcnh committed:
Merge pull request #2 from fhoeben/javascriptSlimExpressions

Javascript slim expressions

tcnh committed:
All wiki-side translations seem to work

fhoeben commented on src/fitnesse/testsystems/slim/tables/SlimTable.java:
```diff
@@ -200,7 +203,28 @@ public void setCustomComparatorRegistry(CustomComparatorRegistry customComparato
+
+    if (getSymbol(variableName) != null) {
+      Map<String, MethodExecutionResult> symbols = new HashMap<>();
+      symbols.put(variableName, new MethodExecutionResult(getSymbol(variableName), Object.class));
```
Shouldn't all symbols from the test's context be made accessible to the evaluator (the expression might reference more than one variable, and might not start with a variable's name). I mean the expression could be `1.5 * myHashMap.size() + myHashMap2.key2.length()`. 
Or am I misinterpreting something?

fhoeben commented on src/fitnesse/slim/SlimExpressionEvaluator.java:
```diff
@@ -1,12 +1,19 @@
-import java.util.Map;
+import java.util.*;
+
+import static org.junit.Assert.assertEquals;
```
This import I was not expecting....

fhoeben commented on src/fitnesse/slim/SlimExpressionEvaluator.java:
```diff
@@ -34,6 +41,9 @@ public void setContext(Map<String, MethodExecutionResult> variables) {
     for (Map.Entry<String, MethodExecutionResult> entry : variables.entrySet()) {
       String key = entry.getKey();
       Object value = entry.getValue().getObject();
+      if(value.toString().startsWith("!{")) {
```
Shouldn't the handling for 'wiki side' expressions be separate from the 'slim server side'?
For slim-server-side I don't expect to ever encounter this prefix, correct?

fhoeben reviewed (commented):
Start is there but....

tcnh committed:
Have all symbols available for the expressionEvaluator
Separate utility to parse wiki symbols to their html representation
Removed unused imports

tcnh commented:
You're very right :)

I've separated the wiki markup parsing using an extra helper, all symbols are now availablet to the expression evaluator (needed an extra method in SlimTestContext to get all symbols). Did some cleanup on the ugly imports..

I think it looks better now

Commit:
Fix \!headings when an heading contains literal macro

tcnh committed:
Updated regex for parsing symbols when evaluating results. Now the diff plays nice with expressions too

tcnh committed:
Also parse symbols in the wiki when 'check not' is called (didn't happen ever)

tcnh committed:
Moved acceptance tests to a better location

tcnh committed:
Sharper regex so that we can concatenate expression results

tcnh committed:
The missing regex change :)

tcnh commented:
Hehe it helps to tick all the boxes when committing. I agree using multiple expressions in one cell is not a good idea.
in-cell concatenating isn't working right now. Needs some attention. I'll see if I can fix that tomorrow

tcnh committed:
Concatenating expressions in strings

tcnh committed:
A unit test for evaluating slim expressions

tcnh committed:
Documentation page for expressions in tables

tcnh commented:
I think it's pretty much complete now. What do you think? 
Did I overlook anything, or is it time to create a new PR on the fitnesse repo?

fhoeben commented on src/fitnesse/slim/SlimExpressionEvaluator.java:
```diff
@@ -5,8 +5,7 @@
 import javax.script.ScriptException;
-import java.util.LinkedHashMap;
-import java.util.Map;
+import java.util.*;
```
The changes in this file do not seem to add much value, correct?

fhoeben commented on src/fitnesse/testsystems/slim/tables/WikiSymbolTranslateUtil.java:
```diff
@@ -0,0 +1,104 @@
+/**
+ * Helper class to translate wiki markup in symbols to a Slim-readable html String
+ */
+
```
Why the empty line?

fhoeben reviewed (approved):
Looks quite good. Ideally I would like the Slim expression evaluation (or any Slim symbol's value actually) to be pushed to the Slim server, so there would be no need to messing around with (possibly outdated) String values on the wiki side. Something to the extend (as adding a command to do `|$var=|clone symbol|$var|`, where var might also be an expression between backticks ``)
I fear that might require changes in the Slim protocol however, because I'm not sure all Slim servers support `clone symbol`. 

Do you have any clue how hard it would be to add an extra slim command (e.g. `|$var=|clone symbol|$var|`) when the server is known to support it?

tcnh commented:
I think the problem is that clone symbol also lives on the slim server side (helper implementing statementExecutionerInterface). And if I read it right, the wiki-table side evaluation happens before anything on the server side is executed.

Life would be so much easier if there was only one administration of slim symbols, but I'm afraid that's not going to happen anytime soon :)

tcnh committed:
Fix velocity templates from jar by setting the current thread's context classloader to Fitnesse's extended classloader containing plugin Jars.

Commit:
#1126 regard StopTest in SuiteSetUp page as StopSuite

briehman committed:
Remove non-standard XML transformer attribute

Removes the attribute intended to specify the indent settings for XML
transformation settings that may not be supported by all transformers.
Specifically, the library xalan-2.7.1 throws an exception when this code
is encountered.

jaruzafa committed:
Revert "Moved plugin declaration"

This reverts commit f97e682566d8c46ef9b2a7147c7f837ed539594b.

jaruzafa committed:
Revert "Added jacoco plugin"

This reverts commit e7ae9aa623493b891fb851188a089f31c811bf47.

jaruzafa committed:
Merge remote-tracking branch 'upstream/master'

jaruzafa committed:
Added test to check that TodaysDate is rendered in Preformatted text blocks

six42 committed:
Allow to test for exceptions

six42 committed:
updated test cases and documentation

six42 committed:
fixed EOL settings to get useful diff

six42 committed:
added configuration switch to disable exception matching

This is mainly intended for backward compatibility

tcnh committed:
Convert lists' string representation to an actual list when evaluating expressions

tcnh committed:
Only treat strings between brakets as list

Commit:
Permit multiline arguments for custom comparators

tcnh committed:
Brace to the right line

amolenaar committed:
Prepare release notes file for a new release

fhoeben committed:
Merge pull request #1127 from tcnh/fixVelocityTemplatesFromJar

Fix velocity templates from jar not loading

fhoeben committed:
Update release notes to describe issue fixed

fhoeben committed:
Fix class name in doc

fhoeben committed:
Merge pull request #1128 from elmuerte/master

resolves #1126: StopTest in SuiteSetUp should be handled as StopSuite

fhoeben committed:
Add test to ensure "StopTestException" is SuiteSetUp is handled as SuiteStopException (#1126)

fhoeben committed:
Merge in changes from #992

fhoeben committed:
Fix unit test for new behavior of stop suite on stop exection in suite setup

fhoeben committed:
Add unit test to check new behavior

fhoeben committed:
Merged #1017

fhoeben committed:
Add PRs merged to release notes

fhoeben committed:
Add test to prevent FITNESSE_VERSION variable not being set (#1170)

fhoeben committed:
Merged in https://github.com/unclebob/fitnesse/pull/1131, ability to check exception

fhoeben committed:
Add to release notes

fhoeben committed:
Merge branch '1126-xml-transformer' of https://github.com/briehman/fitnesse into briehman-1126-xml-transformer

fhoeben committed:
Add to release notes

fhoeben committed:
Merge pull request #1142 from dcutting/master

Permit multiline arguments for custom comparators

fhoeben committed:
Add to release notes

fhoeben committed:
Merge branch 'fix/IdInHeadings' of https://github.com/tavsccn/fitnesse into tavsccn-fix/IdInHeadings

fhoeben committed:
Add to release notes

fhoeben committed:
Merge pull request #1018 from pepeZpew/master

replace ConcurrentHashMap with plain HashMap to handle null keys

fhoeben committed:
Also add .wiki files to jars

fhoeben committed:
Add malformed request to exception message

fhoeben committed:
Ignore empty lines before request lines, fixes #1042

fhoeben committed:
Add to release notes

fhoeben committed:
Merge pull request #1172 from fhoeben/includeWikiInJar

Also add .wiki files to jars

fhoeben committed:
Fix #1173, Prevent endless loop when browser keeps sending empty lines.

fhoeben committed:
Ensure we remove ExecutionException which wraps the actual exception thrown, so that catch clauses present actually work as intended

fhoeben committed:
Add to release notes

fhoeben committed:
Prepare for next release

tcnh committed:
Make request data available to responders (#1174)

* Make request data available to responders

fhoeben committed:
Add #1174 to release notes

achmyr committed:
Adding support for Android slim client (#602)

Adding support for standalone slim client that can be used by Android app.

fhoeben committed:
Prepare for next developments

fhoeben committed:
Reduce size of .jar by not including all classes twice...

fhoeben committed:
Upgrade to gradle 5.1

fhoeben committed:
Update release notes

fhoeben committed:
Fix handling of decision table with unknown output parameter (#1178)

* Throw similar error for incorrect output as input parameter for decision table based on scenario

fhoeben committed:
Latest bintray plugin

fhoeben committed:
Prepare for next developments

fhoeben committed:
Dependency updates and compile target set to Java 8 (#1176)

* Target Java 8
* Upgrade dependencies
* Can we build with openjdk also?

fhoeben committed:
Prepare for next developments

fhoeben committed:
Merge branch 'master' into referenceMapValuesFromSlimSymbols

fhoeben committed:
Add to release notes

fhoeben committed:
Ensure processing of lists also works if custom converter is registered for lists

fhoeben committed:
Further fix

fhoeben committed:
Make test independent on whether another test previously registered a custom converter for lists

fhoeben committed:
Simplify conversion of lists, also allow empty lists

tcnh committed:
Merge pull request #3 from fhoeben/referenceMapValuesFromSlimSymbols

Reference map values from slim symbols
